rskj-core/src/main/java/org/ethereum/core/BlockHeader.java

```diff
@@ -318,20 +291,28 @@ public Keccak256 getHash() {
     public byte[] getFullEncoded() {
         // the encoded block header must include all fields, even the bitcoin PMT and coinbase which are not used for
         // calculating RSKIP92 block hashes
-        return this.getEncoded(true, true);
+        return this.getEncoded(true, true, false);
     }
 
     public byte[] getEncoded() {
         // the encoded block header used for calculating block hashes including RSKIP92
-        return this.getEncoded(true, !useRskip92Encoding);
+        return this.getEncoded(true, !useRskip92Encoding, false);
+    }
+
+    public byte[] getEncodedForHeaderMessage() {
+        return this.getEncoded(true, true, true);
+    }
+
+    public byte[] getEncodedForHash() {
+        return this.getEncoded(true, !useRskip92Encoding, true);
     }
 
     @Nullable
     public Coin getMinimumGasPrice() {
         return this.minimumGasPrice;
     }
 
-    public byte[] getEncoded(boolean withMergedMiningFields, boolean withMerkleProofAndCoinbase) {
```

Comment:
As there are two versions so far, could we use the RSKIPs to differentiate between versions? Also, if we have 5 different versions, there'll be 5 classes

Reply:
- As there are two versions so far, could we use the RSKIPs to differentiate between versions?
  I tried to use `if (RSKIP351 is active) newBehavior()` like because of two things: it prevents adding if clause each time the new behavior is needed + net module does not have easy access to activation config. So the result is that block Factory will set the behavior when creating the block instance. It's kind of null object pattern
- Also, if we have 5 different versions, there'll be 5 classes.
  That is the idea. if we have 5 different versions we avoid tons of `if (RSKIP351 is active) newBehavior()`